docs/proposal/2022-09-28-event-manager-extensibility.md

```diff
@@ -0,0 +1,302 @@
+
+- [ ] Add plugin contribution documentation
+  - Prepare a documentation for adding new plugin to Botkube Plugins.
+  - It should contain how to verify plugin functionality in PR checks.
+  - Boilerplate template for plugin development. Please refer [here](https://github.com/huseyinbabal/botkube-plugins/tree/main/plugins/kubectl) for an example.
+  - Once needed, provide a mocking mechanism to write tests for specific plugin.
+  
+- [ ] Add plugin manager to Botkube
+  - Plugin manager can accept plugin configuration with name and version.
+  - It should download plugin executable by using name and version to specified local folder before running Botkube application
+  - There should be a caching mechanism to not download same version again.
+  - You can see working example [here](https://github.com/huseyinbabal/botkube-plugins-playground/blob/main/plugin/manager.go)
+
+- [ ] Extract Kubernetes source feature as a plugin 
+  - Currently, Kubernetes source events are tightly coupled with Botkube, we can add sourcing mechanism as `kubernetes` plugin. You can see independent example [here](https://github.com/huseyinbabal/botkube-plugins/tree/main/plugins/kubernetes)
+  - It should be background compatible.
+  - Plugin specific parameters should be passed during plugin initialization. Hashicorp's Go plugin system, runs plugin executable as sub process, so it also accepts exec arguments where we can resolve configuration and pass them as separate arguments
+  - Plugin parameters will be documented in [botkube-plugins](https://github.com/huseyinbabal/botkube-plugins) repo, and they can be provided from Botkube as an environment variable like `KUBERNETES_...` so that they can be passed to executable [in this line](https://github.com/huseyinbabal/botkube-plugins-playground/blob/c85cb7b84296a2f41c2dcdd8cac77c0e9dd9c69a/plugin/manager.go#L159)
+  - Data structure for events
+    - Since each source plugin has different data structure, we cannot have a general data structure except we specified #Cloud Events. `data` field has free data structure, but we can convert them to map then extract 
+any section based on customer definition by using Go template. `data` field has free structure, but we can put mandatory fields in this struct like `message`, `user`, etc. Or we can put mandatory fields to cloud events schema as shown below.
+      ```json
+      {
+      "id" : "e2361318-e50b-472e-8b17-13dbac1daac1",   
+      "source" : "Kubernetes",     
+      "specversion" : "1.0",                             
+      "time" : "2022-03-23T14:35:39.738Z",               
+      "message": "DB instance not found",
+      "user": "john",
+      "data" : {                                         
+        "namespace" : "Botkube",
+          "deployment" : {
+            "name" : "test",
+            "replicas" : 5
+          }
+        }
+      }
+      ```
+- [ ] Extract Kubectl executor feature as a plugin
```

Comment:
Will we give them a proper rights? How you see extracting the `kubectl` to a dedicated plugin when it comes to the `commands.verb` and `commands.resources` configuration?
Will be good to establish some common agreement about short and long term solution before jumping into implementation.

Reply:
I am not sure if I understand correctly, but when we extract to plugin, there will be no difference between current format and pluggable format. Configurations are blindly passed by plugin manager, but the actual validation and process should be handle by plugin. If conf is incorrect, then once Botkube calls this plugin, it will receive error response and we can handle exceptions.

Reply:
This is about RBAC, so Botkube currently has a cluster-admin role and kubectl executor simpily uses that but on the top of that we perform some validation to ensure that it's aligned with the permission defined in the configuration

Reply:
Ok, you are right, we cannot use multiple service accounts in one pod that means we cannot use different service accounts for kubectl and helm. However, I cannot see any concern about that in the beginning since;
- Think about your workstation machine that you use kubectl. Once you install helm, you don't really try to think which service account you should use. So you have unified service account and some roles are attached to it, you use it for both helm and kubectl
- Botkube is like a workstation machine where you virtually jump and execute command there. We can think same for Botkube, and customer designs a unified service account with respective roles then attach it to Botkube deployment.
- If we got a feedback about this, then we might think to come up pod-per-plugin notation, which would have operational cost at the beginning.
wdyt?